client/js/color.js

```diff
@@ -0,0 +1,103 @@
+export function toHex(inputColor) {
+  const hexPattern = /^#([A-Fa-f0-9]{6})$/;
+  if (hexPattern.test(inputColor)) {    
+    return inputColor;
+  }
+  const canvas = document.createElement('canvas');
+  const ctx = canvas.getContext('2d');
+  ctx.fillStyle = inputColor;
+  const hexColor = ctx.fillStyle;  
+  canvas.remove();
+  return hexPattern.test(hexColor) ? hexColor : '#000000';
+}
+
+export function toRGB(inputColor) {
+  let hex = toHex(inputColor);
+  let r = parseInt(hex.slice(1, 3), 16);
+  let g = parseInt(hex.slice(3, 5), 16);
+  let b = parseInt(hex.slice(5, 7), 16);
+  return `rgb(${r},${g},${b})`; 
+}
+
+export function calcLuminance(inputColor) {
+  // Function created by https://chat.openai.com/chat
+  let hex = toHex(inputColor);
+  let rgbString = toRGB(hex);
+  let [r, g, b] = rgbString.match(/\d+/g).map(str => parseInt(str, 10));
+  let RsRGB = r / 255;
+  let GsRGB = g / 255;
+  let BsRGB = b / 255;
+  let R, G, B;
+  function applyGammaCorrection(value) {
+    if (value <= 0.04045) {
+      return value / 12.92;
+    } else {
+      return Math.pow((value + 0.055) / 1.055, 2.4);
+    }
+  }
+  R = applyGammaCorrection(RsRGB);
+  G = applyGammaCorrection(GsRGB);
+  B = applyGammaCorrection(BsRGB);
+  const luminance = (0.2126 * R) + (0.7152 * G) + (0.0722 * B);
+  return luminance;
+}
+
+export function calcContrast(color1, color2) {
+  let lum1 = calcLuminance(color1);
+  let lum2 = calcLuminance(color2);
+  return (Math.max(lum1,lum2)+0.05) / (Math.min(lum1,lum2)+0.05);  
+}
+
+export function contrastAnyColor(inputColor, intensity) {
+  // Function created by https://chat.openai.com/chat
+  let color = toHex(inputColor);
+  const luminance = calcLuminance(color);
+  if (luminance > 0.1791) intensity = -intensity;
+  if (intensity == 1)
+    return '#ffffff';
+  else if (intensity == -1)
+    return '#000000';
+  const r = parseInt(color.slice(1, 3), 16);
+  const g = parseInt(color.slice(3, 5), 16);
+  const b = parseInt(color.slice(5, 7), 16);
+  const adjustColor = (r, g, b, intensity) => [r, g, b].map(c => Math.round(Math.max(0, Math.min(c + intensity * 255, 255))));
+  const toHexString = (r, g, b) => '#' + r.toString(16).padStart(2, '0') + g.toString(16).padStart(2, '0') + b.toString(16).padStart(2, '0');
```

Comment:
Splitting `toRGB` into two functions would simplify this line to something like `const toHexString = toHex(rgbArrayToString([r,g,b]))`.